pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -416,14 +416,27 @@ func (c *Cluster) startOSDDaemonsOnPVC(pvcName string, config *provisionConfig,
 			config.addError(errMsg)
 			continue
 		}
-		_, err = c.context.Clientset.AppsV1().Deployments(c.Namespace).Create(dp)
-		if err != nil {
-			if !errors.IsAlreadyExists(err) {
+
+		createdDeployment, createErr := c.context.Clientset.AppsV1().Deployments(c.Namespace).Create(dp)
+		if createErr != nil {
+			if !errors.IsAlreadyExists(createErr) {
 				// we failed to create job, update the orchestration status for this pvc
-				logger.Warningf("failed to create osd deployment for pvc %s, osd %v: %+v", osdProps.pvc.ClaimName, osd, err)
+				logger.Warningf("failed to create osd deployment for pvc %s, osd %v: %+v", osdProps.pvc.ClaimName, osd, createErr)
 				continue
 			}
 			logger.Infof("deployment for osd %d already exists. updating if needed", osd.ID)
+			createdDeployment, err = c.context.Clientset.AppsV1().Deployments(c.Namespace).Get(dp.Name, metav1.GetOptions{})
+			if err != nil {
+				logger.Warningf("failed to get existing OSD deployment %s for update: %+v", dp.Name, err)
+				continue
+			}
+		}
+
+		// keyring must be generated before update-and-wait since no keyring will prevent the
+		// deployment from reaching ready state
+		c.generateKeyring(strconv.Itoa(osd.ID), createdDeployment)
```

Comment:
This pattern of generating the keyring after the deployment is created has a problem that I've seen on the other daemons. There is a race condition that may result in the pod failing to start on the first attempt because the secret doesn't exist yet. While the pod will start successfully on the next attempt, it does make the advanced user wonder why there was even a temporary failure. 

If we're just doing this in order to assign the owner reference to the deployment, I'd suggest we create the secret first, then update the secret with the owner reference if needed. Updating the secret is less intrusive than having the pod fail and restart. But this could be done in a separate PR since other daemons are also affected by the pattern.

Reply:
I have noticed the race and considered it a necessary evil, but I hadn't considered that pattern. Yes, we should update the other daemons with that later.